### PR TITLE
Support lingvo DSL dictionary with UTF-8

### PR DESCRIPTION
**Pull request by miurahr**
Some dictionary use .dsl.dz and UTF-8 encoding for text.
It support such dictionary.

[RFE #1530](https://sourceforge.net/p/omegat/feature-requests/1530/) 

Signed-off-by: Hiroshi Miura <miurahr@linux.com>

### CODE REVIEW

File: src/org/omegat/core/dictionaries/LingvoDSL.java [hunk 1]
```diff
@@ -78,17 +79,34 @@ public IDictionary loadDict(File file, Language language) throws Exception {
...
+                try (InputStreamReader isr = new InputStreamReader(bis, StandardCharsets.UTF_16);
+                     BufferedReader reader = new BufferedReader(isr)) {
+                    loadData(reader);
+                }
             }
         }
 
-        private void loadData(Stream<String> stream) {
+        private void loadData(final BufferedReader reader) {
+            Stream<String> stream = reader.lines();
```
Comment (miurahr) [inline src/org/omegat/core/dictionaries/LingvoDSL.java:109]: No need to change interface here.

File: src/org/omegat/core/dictionaries/LingvoDSL.java [hunk 2]
```diff
@@ -78,17 +79,34 @@ public IDictionary loadDict(File file, Language language) throws Exception {
...
+                try (InputStreamReader isr = new InputStreamReader(bis, StandardCharsets.UTF_8);
+                     BufferedReader reader = new BufferedReader(isr)) {
+                    loadData(reader);
+                }
+            } else {
+                try (InputStreamReader isr = new InputStreamReader(bis, StandardCharsets.UTF_16);
+                     BufferedReader reader = new BufferedReader(isr)) {
+                    loadData(reader);
+                }
             }
```
Comment (amake) [inline src/org/omegat/core/dictionaries/LingvoDSL.java:105]:
```suggestion
            Charset charset = bis.hasBOM() ? StandardCharsets.UTF_8 : StandardCharsets.UTF_16;
            try (InputStreamReader isr = new InputStreamReader(bis, charset);
                 BufferedReader reader = new BufferedReader(isr)) {
                loadData(reader);
            }
```

File: src/org/omegat/core/dictionaries/LingvoDSL.java [hunk 3]
```diff
@@ -29,16 +29,17 @@
...
 import java.nio.charset.StandardCharsets;
-import java.nio.file.Files;
 import java.util.List;
 import java.util.Locale;
 import java.util.regex.Pattern;
 import java.util.stream.Collectors;
 import java.util.stream.Stream;
 import java.util.zip.GZIPInputStream;
 
+import org.apache.commons.io.input.BOMInputStream;
```
Comment (amake) [inline src/org/omegat/core/dictionaries/LingvoDSL.java:42]: This is a nice find. There's a bunch of hand-written BOM detection in OmegaT that we could replace with this.

File: src/org/omegat/core/dictionaries/LingvoDSL.java [hunk 4]
```diff
@@ -78,17 +79,34 @@ public IDictionary loadDict(File file, Language language) throws Exception {
...
+                }
             }
         }
 
-        private void loadData(Stream<String> stream) {
+        private void loadData(final BufferedReader reader) {
+            Stream<String> stream = reader.lines();
             StringBuilder word = new StringBuilder();
             StringBuilder trans = new StringBuilder();
             stream.filter(line -> !line.isEmpty() && !line.startsWith("#"))
```
Comment (amake) [inline src/org/omegat/core/dictionaries/LingvoDSL.java:112]:
```suggestion
            StringBuilder word = new StringBuilder();
            StringBuilder trans = new StringBuilder();
            reader.lines().filter(line -> !line.isEmpty() && !line.startsWith("#"))
```

**Review by amake — COMMENTED** [review]
Did you really mean to delete the `test.dsl.dz` file?

File: src/org/omegat/core/dictionaries/LingvoDSL.java (hunk 3 above)
Comment (amake) [reply src/org/omegat/core/dictionaries/LingvoDSL.java:42]: (I'm not suggesting you do that in this PR.)

File: src/org/omegat/core/dictionaries/LingvoDSL.java (hunk 2 above)
Comment (miurahr) [reply src/org/omegat/core/dictionaries/LingvoDSL.java]: Accept suggestion.

**Comment by miurahr** [comment]
> Did you really mean to delete the `test.dsl.dz` file?

No, I'm doing

```
iconv  -f utf-16le -t utf-8 dict-lingvo/test.dsl dict-lingvo-dz/test.dsl
dictzip dict-lingvo-dz/test.dsl
```

Test data is written only in Latin characters, so UTF16LE -> UTF8  reduce the size half.

**Comment by amake** [comment]
> Test data is written only in Latin characters, so UTF16LE -> UTF8 reduce the size half.

Sorry, I misunderstood the GitHub diff display: it said "-45 bytes" in red, and I thought you had deleted the entire file.

Actually if you're OK with it, I'd like to keep the old file. The reason is that git keeps all files ever added, so this change doesn't actually save any space in the repository. Instead we now have (size of old file) + (size of new file).

**Comment by amake** [comment]
I've made some minor adjustments and committed to master.

Thanks again for your contributions!

**Pull request closed without merging**